paddle/framework/operator.cc

```diff
@@ -411,7 +411,34 @@ void OperatorWithKernel::Run(const Scope& scope,
                  expected_kernel_key);
   }
 
-  kernel_iter->second->Compute(ctx);
+  if (actual_kernel_key == expected_kernel_key) {
+    kernel_iter->second->Compute(ctx);
+  } else {
+    Scope& op_scope = scope.NewScope();
```

Comment:
@reyoung @dzhwinter @jacquesqiao I find that we can not cache the transformed result variables in current scope in order to reduce the transform times. Following is an example:
```
       / op2
op1 ---
       \ op3
```

The output of op1 is the input of op2 and op3.

If we make cache in current scope,

- In the first batch training:
op2 runs first and creates a new variable (var_name + KernelType) and make data transform.
Then, op3 will check if this variable has been created or not. Since this new variable has been created by op2, op3 will directly use it and has no need to make data transform.

- In the second batch training:
We have to make data transform again. But we still check if the new variable is created, the data transform will be skipped.

Reply:
I check the Executor, in every batch, the local scope will be deleted. So this problem will not happen. I will change the cache to local scope instead of op scope.

Reply:
since each batch will create a new local_scope, add a  cache seems can work for our framework.